aten/src/ATen/PythonTorchFunctionTLS.cpp

```diff
@@ -18,6 +18,26 @@ void PythonTorchFunctionTLS::swap_mode(std::shared_ptr<c10::SafePyObject>& mode)
   pythonTorchFunctionState.mode_.swap(mode);
 }
 
+void PythonTorchFunctionTLS::push_onto_stack(std::shared_ptr<SafePyObject> mode) {
+  pythonTorchFunctionState.stack_.push_back(std::move(mode));
+}
+
+const std::shared_ptr<SafePyObject>& PythonTorchFunctionTLS::pop_stack() {
+  TORCH_CHECK(pythonTorchFunctionState.stack_.size() > 0, "trying to pop from empty mode stack");
+  const auto out = pythonTorchFunctionState.stack_.back();
+  pythonTorchFunctionState.stack_.pop_back();
+  return std::move(out);
+}
+
+const std::shared_ptr<SafePyObject>& PythonTorchFunctionTLS::get_stack_at(int64_t idx) {
+  TORCH_CHECK(idx < static_cast<int64_t>(pythonTorchFunctionState.stack_.size()), "Tried to get stack at idx that's too big");
+  return pythonTorchFunctionState.stack_[idx];
+}
+
+int64_t PythonTorchFunctionTLS::stack_len() {
+  return pythonTorchFunctionState.stack_.size();
+}
```

Comment:
This is the big thing that changed for the modes--the TLS now holds the C++ mode (which is the stack running mode or nothing) and the stack, which is a vector of `SafePyObject`s. This therefore exposed push, pop, get_at(i), and len functions for the stack